engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala

```diff
@@ -125,31 +126,83 @@ object WorkflowActor {
+          case _ => false
+        }
+      }
+
+      val restartOrResumeAttempts = runningExecutions map { case (execution, executionInfos) =>
+        val infos = executionInfosToMap(executionInfos)
+        execution.toBackendCallKey(actor.workflow.namespace) match {
+          case Success(backendCallKey) =>
+            if (actor.backend.isResumable(backendCallKey, infos)) {
+              Future.successful(())
+            } else if (isRunningCollector(execution) || actor.backend.isRestartable(backendCallKey, infos)) {
+              globalDataAccess.updateStatus(actor.workflow.id, Seq(execution.toKey), ExecutionStatus.NotStarted)
+            } else {
+              Future.failed(new Exception(s"Execution ${execution.executionId} is neither restartable or resumable"))
+            }
+          case Failure(ex) => Future.failed(ex)
+        }
+      }
+
+      Future.sequence(restartOrResumeAttempts).map(_ => ())
+    }
+
+    private def filterResumableExecutions(actor: WorkflowActor, runningExecutions: Map[Execution, Seq[ExecutionInfo]]): Future[Map[Execution, Seq[ExecutionInfo]]] = {
+      val x = runningExecutions map { case (execution, executionInfos) =>
+        val infos = executionInfosToMap(executionInfos)
+        execution.toBackendCallKey(actor.workflow.namespace) match {
+          case Success(backendCallKey) if actor.backend.isResumable(backendCallKey, infos) =>
+            Future.successful(Option(execution -> executionInfos))
+          case Success(_) => Future.successful(None)
+          case Failure(ex) => Future.failed(ex)
+        }
+      }
+      Future.sequence(x).map(_.flatten).map(_.toMap)
+    }
+
     override def runInitialization(actor: WorkflowActor): Future[Unit] = {
+      def isWorkflowRestartable(allExecutions: Traversable[Execution]): Future[Unit] = {
+        allExecutions.filter(e => Seq(ExecutionStatus.Aborted, ExecutionStatus.Failed).contains(e.executionStatus)) match {
+          case es if es.nonEmpty =>
+            val string = es.toSeq.sortWith((lt, rt) => lt.callFqn < rt.callFqn || (lt.callFqn == rt.callFqn && lt.index < rt.index)).mkString(" ")
+            Future.failed(new Throwable(s"Workflow ${actor.workflow.id.shortString}: Cannot restart, found Failed and/or Aborted executions: $string"))
+          case _ => Future.successful(())
+        }
+      }
+
       for {
-        _ <- actor.backend.prepareForRestart(actor.workflow)
-        _ <- actor.dumpTables()
+        allExecutions <- globalDataAccess.getExecutions(actor.workflow.id)
+        _ <- isWorkflowRestartable(allExecutions)
+        runningExecutionsAndInfos <- globalDataAccess.runningExecutionsAndExecutionInfos(actor.workflow.id)
+        executionsAndInfosMap = runningExecutionsAndInfos.map(x => x.execution -> x.executionInfos).toMap
+        _ <- restartOrResume(actor, executionsAndInfosMap)
```

Comment:
Everything added above here was moved here [from the JesBackend](https://github.com/broadinstitute/cromwell/pull/704/files#diff-02fb95673814eb1a1aca084e078b5113L795).  It's definitely messy and even I have a hard time following it... I'll try to clean it up but any suggestions would be welcome